File: typescript/Structural/09-Decorator/Decorator/JobCardDecorator.ts
import { IJobCard } from "../Component/IJobCard";

// Wraps an IJobCard and delegates all calls to it by default.
// Concrete decorators only override what they need to change.
export abstract class JobCardDecorator implements IJobCard {
  constructor(protected readonly wrapped: IJobCard) {}

  get jobId(): string {
    return this.wrapped.jobId;
  }
  get title(): string {
    return this.wrapped.title;
  }
  get category(): string {
    return this.wrapped.category;
  }
  get cost(): number {
    return this.wrapped.cost;
  }
  get description(): string {
    return this.wrapped.description;
  }

  execute(): void {
    this.wrapped.execute();
  }
}

File: csharp/Structural/09-Decorator/Decorator/JobCardDecorator.cs
using Decorator.Component;

namespace Decorator.Decorator;

// Wraps an IJobCard and delegates all calls to it by default.
// Concrete decorators only override what they need to change.
public abstract class JobCardDecorator : IJobCard
{
    protected readonly IJobCard _wrapped;

    protected JobCardDecorator(IJobCard jobCard)
    {
        _wrapped = jobCard;
    }

    public virtual string JobId => _wrapped.JobId;
    public virtual string Title => _wrapped.Title;
    public virtual string Category => _wrapped.Category;
    public virtual decimal Cost => _wrapped.Cost;
    public virtual string Description => _wrapped.Description;
    public virtual void Execute() => _wrapped.Execute();
}
